File: Daily/2024-04/2024-04-09/31741.구간 덮기.py
import sys; input = sys.stdin.readline
n, l, r = map(int, input().split())
arr = [[],[],[]]
ans = int(1e10)
for _ in range(n):
    a,b = map(int, input().split())
    if a <= l and b >= r:
        print(0)
        exit()
    if b < l or a > r:
        continue
    if a <= l:
        arr[0].append(b)
    elif b >= r:
        arr[2].append(a)
    else:
        arr[1].append((a,b))
if not arr[0] or not arr[2]:
    print(-1)
    exit()
arr[0].sort()
arr[2].sort()

def bisect1(y):
    global ans
    st = 0
    fi = len(arr[2]) - 1
    res = 0
    while st <= fi:
        mid = (st + fi) // 2
        if arr[2][mid]> y:
            fi = mid - 1
        else:
            st = mid + 1
            res = arr[2][mid]
    if not res:
        return
    ans = min(ans, abs(y - res))
def bisect2(x,y):
    global ans
    st = 0
    fi = len(arr[0]) - 1
    res = 0
    tem = 0
    while st <= fi:
        mid = (st + fi) // 2
        if arr[0][mid] < x:
            st = mid + 1
        else:
            fi = mid - 1
            res = arr[0][mid]
    if not res:
        return
    tem += abs(res - x)
    st = 0
    fi = len(arr[2]) - 1
    res = 0
    while st <= fi:
        mid = (st + fi) // 2
        if arr[2][mid] > y:
            fi = mid - 1
        else:
            st = mid + 1
            res = arr[2][mid]
    if not res:
        return
    tem += abs(y - res)
    ans = min(ans, tem)

for x in arr[0]:
    bisect1(x)
for x,y in arr[1]:
    bisect2(x,y)
ans = -1 if ans == int(1e10) else ans
print(ans)

File: Daily/2024-04/2024-04-09/31741.구간 덮기.rs
use std::{io, vec};
use std::io::prelude::*;

fn read<T>(si: &mut T) -> String where T: Read {
    let mut s = String::new();
    si.read_to_string(&mut s).unwrap();
    s
}

fn next<T>(it: &mut std::str::SplitAsciiWhitespace) -> T where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug {
    it.next().unwrap().parse().unwrap()
}

fn bisect(arr: &Vec<(i64,i64)>, A_arr: &Vec<i64>, B_arr: &Vec<i64>, l: i64, r: i64) -> i64 {
    let mut st: i32 = 0;
    let mut fi:i32 = A_arr.len() as i32 - 1;
    let mut res = 0;
    let mut ans = 0;
    while st <= fi {
        let mid = (st+fi) >> 1;
        if A_arr[mid as usize] < l {
            st = mid + 1;
        } else {
            fi = mid - 1;
            res = A_arr[mid as usize] ;
        }
    }
    if res == 0  {
        return i64::MAX;
    }
    ans += i64::abs(res - l);
    res = 0;
    st = 0;
    fi = B_arr.len() as i32 - 1;
    while st <= fi {
        let mid = (st+fi) >> 1;
        if B_arr[mid as usize] > r {
            fi = mid - 1;
        } else {
            st = mid + 1;
            res = B_arr[mid as usize] ;
        }
    }
    if res == 0 {
        return i64::MAX;
    }
    ans += i64::abs(r - res);
    ans
}

fn bisect_mid(B_arr: &Vec<i64>, r: i64) -> i64 {
    let mut st: i32 = 0;
    let mut fi:i32 = B_arr.len() as i32 - 1;
    let mut res = 0;
    while st <= fi {
        let mid = (st+fi) >> 1;
        if B_arr[mid as usize] > r {
            fi = mid - 1;
        } else {
            st = mid + 1;
            res = B_arr[mid as usize] ;
        }
    }
    if res == 0 {
        return i64::MAX;
    }
    i64::abs(r - res)
}

fn main() {
    // 입력
    let mut si = io::BufReader::new(io::stdin().lock());
    let mut so = io::BufWriter::new(io::stdout().lock());
    let s = read(&mut si);
    let mut it = s.split_ascii_whitespace();

    let n = next::<i64>(&mut it);
    let l = next::<i64>(&mut it);
    let r = next::<i64>(&mut it);

    let mut ans = i64::MAX;
    let mut arr: Vec<(i64,i64)> = Vec::new();
    let mut A_arr = Vec::new();
    let mut B_arr = Vec::new();
    for _ in 0..n {
        let a = next::<i64>(&mut it);
        let b = next::<i64>(&mut it);
        if a <= l && b >= r {
            ans = 0;
        }
        if b < l || a > r {
            continue;
        }
        if a <= l {
            A_arr.push(b);
        } else if b >= r {
            B_arr.push(a);
        } else {
            arr.push((a, b));
        }
    }
    if ans == 0 {
        writeln!(so, "0").ok();
    } else {
        A_arr.sort_unstable();
        B_arr.sort_unstable();
        for i in 0..arr.len() {
            let (x,y) = arr[i];
            ans = ans.min(bisect(&arr,&A_arr,&B_arr,x, y));
        }
        for i in 0..A_arr.len() {
            let x = A_arr[i];
            ans = ans.min(bisect_mid(&B_arr, x));
        }
        if ans == i64::MAX {
            writeln!(so, "-1").ok();
        } else {
            writeln!(so, "{}", ans).ok();
        }
    }
}